plugin/plugin.go

```diff
@@ -0,0 +1,351 @@
+package plugin
+
+import (
+	"context"
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Constants for threshold modes
+const (
+	ThresholdModeAbsolute   = "absolute"
+	ThresholdModePercentage = "percentage"
+	DefaultThresholdMode    = ThresholdModeAbsolute // Default value
+)
+
+// Args represents the plugin's configurable arguments.
+type Args struct {
+	ReportFilenamePattern     string `envconfig:"PLUGIN_REPORT_FILENAME_PATTERN"`
+	FailedFails               int    `envconfig:"PLUGIN_FAILED_FAILS"`
+	FailedSkips               int    `envconfig:"PLUGIN_FAILED_SKIPS"`
+	FailureOnFailedTestConfig bool   `envconfig:"PLUGIN_FAILURE_ON_FAILED_TEST_CONFIG"`
+	ThresholdMode             string `envconfig:"PLUGIN_THRESHOLD_MODE"`
+	Level                     string `envconfig:"PLUGIN_LOG_LEVEL"`
+}
+
+// ValidateInputs ensures the user inputs meet the plugin requirements.
+func ValidateInputs(args Args) error {
+	if args.ReportFilenamePattern == "" {
+		return errors.New("missing required parameter: ReportFilenamePattern. Please specify the pattern to locate the TestNG report files")
+	}
+
+	if args.FailedFails < 0 || args.FailedSkips < 0 {
+		return errors.New("threshold values must be non-negative. Check the configured values for failed and skipped tests")
+	}
+
+	if args.ThresholdMode == "" {
+		args.ThresholdMode = DefaultThresholdMode
+		logrus.Infof("PLUGIN_THRESHOLD_MODE not specified. Defaulting to '%s'", DefaultThresholdMode)
+	} else if args.ThresholdMode != ThresholdModeAbsolute && args.ThresholdMode != ThresholdModePercentage {
+		return errors.New("invalid ThresholdMode value. It must be 'absolute' or 'percentage'. Check the configuration")
+	}
+
+	return nil
+}
+
+// Exec handles TestNG XML report processing and logs details.
+func Exec(ctx context.Context, args Args) error {
+	files, err := locateFiles(args.ReportFilenamePattern)
+	if err != nil {
+		logger := logrus.WithError(err)
+		logger.Error("Error locating files")
+		return errors.New("failed to locate files: " + err.Error())
+	}
+
+	if len(files) == 0 {
+		return errors.New("no TestNG XML report files found. Check the report file pattern")
+	}
+
+	var aggregatedResults Results
+	var skippedFiles []string
+
+	for _, file := range files {
+		results, err := processFile(file)
+		if err != nil {
+			logrus.WithField("File", file).WithError(err).Warn("Skipping file due to errors")
+			skippedFiles = append(skippedFiles, file)
+			continue // Skip this file and process others
+		}
+		aggregatedResults.Total += results.Total
+		aggregatedResults.Failures += results.Failures
+		aggregatedResults.Skipped += results.Skipped
+		aggregatedResults.DurationMS += results.DurationMS
+	}
```

Comment:
```suggestion
        var (
                resultsChan    = make(chan Results, len(files))
                errorsChan     = make(chan error, len(files))
        )
        
        for _, file := range files {
	        go func(f string) {
		        res, err := processFile(f)
		        if err != nil {
			        errorsChan <- fmt.Errorf("failed to process file %s: %w", f, err)
			        return
		        }
		        resultsChan <- res
	        }(file)
        }
        
	var aggregatedResults Results
	var skippedFiles []string
	
	for i := 0; i < len(files); i++ {
	        select {
	        case res := <-resultsChan:
		        aggregatedResults.Total += res.Total
		        aggregatedResults.Failures += res.Failures
		        aggregatedResults.Skipped += res.Skipped
		        aggregatedResults.DurationMS += res.DurationMS
	        case err := <-errorsChan:
		        logrus.Warn(err)
			if e, ok := err.(*os.PathError); ok {
			        skippedFiles = append(skippedFiles, e.Path)
			}
	        }
	}
```

let's implement parallel processing in case of large number of files

Reply:
also add tests with a mix of valid and invalid files in a parallel processing scenario